tokio/examples/chat.rs

```diff
@@ -0,0 +1,257 @@
+                peer.1.send(message.into()).await?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Peer {
+    /// Create a new instance of `Peer`.
+    async fn new(
+        state: Arc<Mutex<Shared>>,
+        lines: Framed<TcpStream, LinesCodec>,
+    ) -> io::Result<Peer> {
+        // Get the client socket address
+        let addr = lines.get_ref().peer_addr()?;
+
+        // Create a channel for this peer
+        let (tx, rx) = mpsc::unbounded();
+
+        // Add an entry for this `Peer` in the shared state map.
+        state.lock().await.peers.insert(addr, tx);
+
+        Ok(Peer { lines, rx })
+    }
+}
+
+#[derive(Debug)]
+enum Message {
+    /// A message that should be broadcasted to others.
+    Broadcast(String),
+
+    /// A message that should be recieved by a client
+    Recieved(String),
+}
+
+// Peer implements `Stream` in a way that polls both the `Rx`, and `Framed` types.
+// A message is produced whenever an event is ready until the `Framed` stream returns `None`.
+impl Stream for Peer {
+    type Item = Result<Message, LinesCodecError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // First poll the `UnboundedReciever`.
+        while let Poll::Ready(Some(v)) = self.rx.poll_next_unpin(cx) {
+            return Poll::Ready(Some(Ok(Message::Recieved(v))));
+        }
+
+        // Secondly poll the `Framed` stream.
+        match self.lines.poll_next_unpin(cx) {
+            Poll::Ready(Some(v)) => {
+                let res = match v {
+                    Ok(msg) => Ok(Message::Broadcast(msg)),
+                    Err(e) => Err(e),
+                };
+                return Poll::Ready(Some(res));
+            }
+            Poll::Ready(None) => {
+                return Poll::Ready(None);
+            }
+            _ => {}
+        }
+
+        // Nothing was ready so we return `Poll::Pending`.
+        Poll::Pending
+    }
+}
+
+/// Process an individual chat client
+async fn process(
+    state: Arc<Mutex<Shared>>,
+    stream: TcpStream,
+    addr: SocketAddr,
+) -> Result<(), Box<dyn Error>> {
+    let mut lines = Framed::new(stream, LinesCodec::new());
+
+    // Read the first line from the `LineCodec` stream to get the username.
+    let username = match lines.next().await {
+        Some(Ok(line)) => line,
+        // We didn't get a line so we return here.
+        _ => return Ok(()),
+    };
+
+    // Register our peer with state which internally sets up some channels.
+    let mut peer = Peer::new(state.clone(), lines).await?;
+
+    // A client has connected, let's let everyone know.
+    {
+        let mut state = state.lock().await;
+        let msg = format!("{} has joined the chat", username);
+        println!("{}", msg);
+        state.broadcast(addr, &msg).await?;
+    }
+
+    // Process incoming messages until our stream is exhausted by a disconnect.
+    while let Some(Ok(msg)) = peer.next().await {
```

Comment:
I wonder if you should just match on the Option, them inner match on the result, this would allow you to matcha `Err(io::Error)` which you can then print to the screen.